src/Tempest/Http/src/Routing/Construction/RouteMatchingRegexesBuilder.php

```diff
@@ -0,0 +1,136 @@
+
+    private const int REGEX_SIZE_MARGIN = 264;
+
+    private const REGEX_SIZE_LIMIT = self::PREG_REGEX_SIZE_LIMIT - self::REGEX_SIZE_MARGIN;
+
+    public function __construct(private RouteTreeNode $rootNode)
+    {
+    }
+
+    public function toRegex(): MatchingRegexes
+    {
+        // Holds all regex "chunks"
+        $regexes = [];
+
+        // Current regex chunk
+        $regex = '';
+        // Used to track how to 'end' a regex chunk partially in the building process
+        $regexBack = '';
+
+        /** @var (RouteTreeNode|null)[] $workingSet */
+        $workingSet = [$this->rootNode];
+
+        // Track how 'deep' we are in the tree to be able to rebuild the regex prefix when chunking
+        /** @var RouteTreeNode[] $stack */
+        $stack = [];
+
+        // Processes the working set until it is empty
+        while ($workingSet !== []) {
+            // Use array_pop for performance reasons, this does mean that the working set works in a fifo order
+            /** @var RouteTreeNode|null $node */
+            $node = array_pop($workingSet);
+
+            // null values are used as an end-marker, if one is found pop the stack and 'close' the regex
+            if ($node === null) {
+                array_pop($stack);
+                $regex .= $regexBack[0];
+
+                $regexBack = substr($regexBack, 1);
+
+                continue;
+            }
+
+            // Checks if the regex is getting to big, and thus if we need to chunk it.
+            if (strlen($regex) > self::REGEX_SIZE_LIMIT) {
+                $regexes[] = '#' . substr($regex, 1) . $regexBack . '#';
+                $regex = '';
+
+                // Rebuild the regex match prefix based on the current visited parent nodes, known as 'the stack'
+                foreach ($stack as $previousNode) {
+                    $regex .= '|' . self::routeNodeSegmentRegex($previousNode);
+                    $regex .= '(?';
+                }
+            }
+
+            // Add the node route segment to the current regex
+            $regex .= '|' . self::routeNodeSegmentRegex($node);
+            $targetRouteRegex = self::routeNodeTargetRegex($node);
+
+            // Check if node has children to ensure we only use branches if the node has children
+            if ($node->dynamicPaths !== [] || $node->staticPaths !== []) {
+                // The regex uses "Branch reset group" to match different available paths.
+                // two available routes /a and /b will create the regex (?|a|b)
+                $regex .= '(?';
+                $regexBack .= ')';
+                $stack[] = $node;
+
+                // Add target route regex as an alteration group
+                if ($targetRouteRegex) {
+                    $regex .= '|' . $targetRouteRegex;
+                }
+
+                // Add an end marker to the working set, this will be processed after the children has been processed
+                $workingSet[] = null;
```

Comment:
I really didn't like using `null` as a meaning-full value . But... When wrapping these things into the smallest possible value-object this whole section became around 60% slower.. I didn't think the readability was worth that much performance loss. (See PR description `RouteMatchingRegexesBuilder performance considerations` for more info)

Reply:
Agree :)